operator/src/main/java/oracle/kubernetes/operator/http/HttpAsyncRequestStep.java

```diff
@@ -0,0 +1,122 @@
+ */
+public class HttpAsyncRequestStep extends Step {
+
+  interface FutureFactory {
+    CompletableFuture<HttpResponse<String>> createFuture(HttpRequest request);
+  }
+
+  private static final LoggingFacade LOGGER = LoggingFactory.getLogger("Operator", "Operator");
+  private static FutureFactory DEFAULT_FACTORY = HttpAsyncRequestStep::createFuture;
+
+  private static final long DEFAULT_TIMEOUT_SECONDS = 5;
+
+  private static FutureFactory factory = DEFAULT_FACTORY;
+  private final HttpRequest request;
+  private long timeoutSeconds = DEFAULT_TIMEOUT_SECONDS;
+
+  private HttpAsyncRequestStep(HttpRequest request, HttpResponseStep responseStep) {
+    super(responseStep);
+    this.request = request;
+  }
+
+  /**
+   * Creates a step to send a GET request to a server. If a response is received, processing
+   * continues with the response step. If none is received within the timeout, the fiber is terminated.
+   * @param url the URL of the targeted server
+   * @param responseStep the step to handle the response
+   * @return a new step to run as part of a fiber, linked to the response step
+   */
+  public static HttpAsyncRequestStep createGetRequest(String url, HttpResponseStep responseStep) {
+    HttpRequest request = HttpRequest.newBuilder(URI.create(url)).GET().build();
+    return create(request, responseStep);
+  }
+
+  /**
+   * Creates a step to send a request to a server. If a response is received, processing
+   * continues with the response step. If none is received within the timeout, the fiber is terminated.
+   * @param request the http request to send
+   * @param responseStep the step to handle the response
+   * @return a new step to run as part of a fiber, linked to the response step
+   */
+  public static HttpAsyncRequestStep create(HttpRequest request, HttpResponseStep responseStep) {
+    return new HttpAsyncRequestStep(request, responseStep);
+  }
+
+  /**
+   * Overrides the default timeout for this request.
+   * @param timeoutSeconds the new timeout, in seconds
+   * @return this step
+   */
+  public HttpAsyncRequestStep withTimeoutSeconds(long timeoutSeconds) {
+    this.timeoutSeconds = timeoutSeconds;
+    return this;
+  }
+
+  @Override
+  public NextAction apply(Packet packet) {
+    AsyncProcessing processing = new AsyncProcessing(packet);
+    return doSuspend(processing::process);
+  }
+
+  class AsyncProcessing {
+    private Packet packet;
+    private CompletableFuture<HttpResponse<String>> future;
+
+    AsyncProcessing(Packet packet) {
+      this.packet = packet;
+    }
+
+    void process(AsyncFiber fiber) {
+      future = factory.createFuture(request);
+      future.whenComplete((response, throwable) -> resume(fiber, response, throwable));
+      fiber.scheduleOnce(timeoutSeconds, TimeUnit.SECONDS, () -> checkTimeout(fiber));
+    }
+
+    private void checkTimeout(AsyncFiber fiber) {
+      if (!future.isDone()) {
+        fiber.terminate(new RuntimeException("timeout"), packet);
```

Comment:
What is the full set of tasks in the steps for this fiber? This says that if the HTTP request times-out, then the full fiber stops.  If this fiber is only reading the status for one server than this is fine. If the fiber is running through a set of servers then none of them will have status updated. If might make more sense to resume the fiber and have the next step (that processes the result) handle that there is no data.

Reply:
Agreed. I will add a commit that changes this behavior.